Considered preferred text output styles for heuristics

doshi-kevin opened the pull request:
<!--
Thanks for contributing to JupyterLab!
Please fill out the following items to submit a pull request.
See the contributing guidelines for more information:
https://github.com/jupyterlab/jupyterlab/blob/main/CONTRIBUTING.md
-->

## References

<!-- Note issue numbers this pull request addresses (should be at least one, see contributing guidelines above). --> #16636 

<!-- Note any other pull requests that address this issue and how this pull request is different. -->

## Code changes

<!-- Describe the code changes and how they address the issue. -->
1. Expanded Supported Output Types
Change:
The list of supported output types has been expanded to include additional MIME types that are commonly used in Jupyter notebooks, specifically:
application/vnd.jupyter.stderr
application/vnd.jupyter.stdout
text/html
image/svg+xml
application/pdf
text/markdown
text
Purpose:
This change ensures that the height estimation considers outputs from standard error and standard output streams, which may contain significant text that contributes to the overall height of the cell. By including these types, the estimation becomes more comprehensive and accurate.
2. Dynamic Output Type Selection
Change:
The code now dynamically selects the preferred output type from the expanded list of supported types using the find method. It checks if the data associated with each output type is a string, allowing for a fallback to the first valid type found.
Purpose:
This approach allows the function to prioritize output types that are likely to contain meaningful text data. By finding the first output type that yields a string, the code ensures that it captures the most relevant output for height estimation, rather than defaulting to a less informative type.
3. Improved Line Counting Logic
Change:
The logic for counting lines in the output data has been refined. The code now checks if the data to display is defined and then counts non-empty lines using a filter.
Purpose:
This enhancement ensures that only meaningful lines are counted, which directly impacts the accuracy of the height estimation. By filtering out empty lines, the function avoids inflating the estimated height with irrelevant data.
4. Total Height Calculation
Change:
The total height is calculated by summing the number of lines in the source code (nLines) and the calculated output lines (outputsLines). This total is then stored in the cellsEstimatedHeight map for future reference.
Purpose:
This final calculation provides a complete and accurate estimation of the widget size. By combining both the source code and relevant output lines, the function delivers a more realistic height for the cell, which is essential for proper rendering in the UI.
How These Changes Address the Issue
The primary issue was that the original height estimation heuristic only considered text/plain outputs, which did not account for other significant output types that could affect the visual representation of the cell. By expanding the supported output types and dynamically selecting the most relevant type, the revised code ensures that all meaningful output is considered in the height estimation.
Benefits of the Changes
Enhanced Accuracy: The new logic captures a broader range of output types, leading to a more accurate estimation of cell height.
Dynamic Adaptability: The function can now adapt to different types of outputs, making it more robust and flexible for various use cases.
Improved User Experience: With better height estimations, the UI can render cells more effectively, enhancing the overall user experience in Jupyter notebooks.
In summary, these changes significantly improve the functionality of the estimateWidgetSize method, ensuring that it provides a more accurate representation of cell heights based on a wider array of output types.

## User-facing changes

<!-- Describe any visual or user interaction changes and how they address the issue. -->

<!-- For visual changes, include before and after screenshots or GIF/mp4/other video demo here. -->

## Backwards-incompatible changes

<!-- Describe any backwards-incompatible changes to JupyterLab public APIs. -->

krassowski commented on packages/notebook/src/windowing.ts:
```diff
@@ -61,18 +61,35 @@ export class NotebookViewModel extends WindowedListModel {
-      NotebookViewModel.DEFAULT_CELL_MARGIN
-    );
+
+    const totalLines = nLines + outputsLines;
+    this.cellsEstimatedHeight.set(model.id, totalLines);
+    return totalLines;
```
This function ought to return an estimated size in pixels, not the number of lines - can you fix it?

krassowski commented on packages/notebook/src/windowing.ts:
```diff
@@ -61,18 +61,35 @@ export class NotebookViewModel extends WindowedListModel {
-    );
+
+    const totalLines = nLines + outputsLines;
+    this.cellsEstimatedHeight.set(model.id, totalLines);
```
This function should not set the estimated height as it is not its responsibility.

doshi-kevin replied:
sure thing, I will make the necessary changes and send another pull request. I am just starting open-source and enjoying this, It would be great to be a contributor to Jupyterlabs.

krassowski replied:
Usually you do not need to open a new pull request, just push changes to this branch and this PR will get updated :)

doshi-kevin replied:
Hey @krassowski , sorry I didn't know about that, I have made the mentioned changes here #16708. Could you please check it and tell me corrections if any.

doshi-kevin replied:
Hey @krassowski any updates on this?
I made a new pull request, could you please check it.

krassowski reviewed (commented):
Thanks for starting this! Leaving some initial feedback inline.

krassowski commented on packages/notebook/src/windowing.ts:
```diff
@@ -52,27 +52,58 @@ export class NotebookViewModel extends WindowedListModel {
   estimateWidgetSize = (index: number): number => {
     const model = this.cells[index].model;
     const height = this.cellsEstimatedHeight.get(model.id);
+
+
```
```suggestion

```

Let's avoid using double new line in side of functions.

krassowski commented on packages/notebook/src/windowing.ts:
```diff
@@ -52,27 +52,58 @@ export class NotebookViewModel extends WindowedListModel {
       return height;
     }
 
+
```
```suggestion
```

krassowski commented on packages/notebook/src/windowing.ts:
```diff
@@ -52,27 +52,58 @@ export class NotebookViewModel extends WindowedListModel {
-    );
+    const totalLines = nLines + outputsLines;
+
+
```
```suggestion
```

krassowski reviewed (commented):
Thank you! Can you add a test case for this as well, following the existing test in:

https://github.com/jupyterlab/jupyterlab/blob/91f1c7c53c78920920a627267dcbae7e9eafed49/packages/notebook/test/modelfactory.spec.ts#L96-L122

As for the code changes description, it would be ok (or even preferred) to write one-two lines summary (for a PR of this ### size). The intent here is to have a high-level overview and call out anything that may need attention (like changes to public API) - it should allow to get a quick understanding of the PR without looking at the source code.

krassowski commented on packages/notebook/src/windowing.ts:
```diff
@@ -52,27 +52,58 @@ export class NotebookViewModel extends WindowedListModel {
+
     if (model instanceof CodeCellModel) {
+      const supportedOutputTypes = [
+        'text/html',
+        'image/svg+xml',
+        'application/pdf',
+        'text/markdown',
+        'text/plain',
+        'application/vnd.jupyter.stderr',
+        'application/vnd.jupyter.stdout',
+        'text'
```
```suggestion
        'text/markdown',
        'text/plain',
        'application/vnd.jupyter.stderr',
        'application/vnd.jupyter.stdout',
        'text'
```

Maybe we should stick to only these as estimates from html/xml code would be likely unreliable?

doshi-kevin replied:
@krassowski  on it

krassowski commented on packages/notebook/src/windowing.ts:
```diff
@@ -52,27 +52,58 @@ export class NotebookViewModel extends WindowedListModel {
     let outputsLines = 0;
+
     if (model instanceof CodeCellModel) {
+      const supportedOutputTypes = [
```
Or maybe not - if we decide to use a more tailored approach as suggested below, then we would not want to sync it

doshi-kevin commented:
Hey @krassowski I have added the test case and removed extra spaces. Kindly check for the same .

krassowski commented on packages/notebook/src/windowing.ts:
```diff
@@ -52,29 +52,91 @@ export class NotebookViewModel extends WindowedListModel {
+}
+
+// Test case for NotebookViewModel
+describe('NotebookViewModel', () => {
```
The test case needs to go to the file I referenced earlier

doshi-kevin commented:
@krassowski  Hey, sorry for inconviniences, I have added that peice of code to modelfactory files and removed extra spaces and edited the code as well.

krassowski commented on packages/notebook/test/modelfactory.spec.ts:
```diff
@@ -152,7 +152,47 @@ describe('@jupyterlab/notebook', () => {
     });
-  });
+
+    test('should calculate height based on number of lines in source and multiple outputs', () => {
```
This test does not actually test the new code added.

Are you using an LLM model? Which one?

doshi-kevin replied:
Yeah for the text model I used the perplexity ai llm.

doshi-kevin replied:
I understood the previous test code and thought that according to the given code added, the test generated by the llm might actually be correct.
Could please tell me where might there be an error so that I can resolve it.

krassowski reviewed (changes requested):
Sorry for taking longer to get back here, busy times. `js-notebook` and `lint, 3.12` jobs are failing on CI which indicates that there are problems with code style and the tests not passing - can you fix these?

doshi-kevin commented:
sure. on it !

doshi-kevin commented:
Hey @krassowski , I have made some linting changes and changes to the test file. Apart from it, I dont understand why (if it still gives error). I will try my best to understand the issue though...

doshi-kevin commented:
@krassowski any updates ?

doshi-kevin commented:
Hey @krassowski , sorry to disturb again, but any updates on this ?

JasonWeill commented:
@doshi-kevin Your tests are still failing; you can see details inline here: https://github.com/jupyterlab/jupyterlab/pull/16706/files#diff-781a703e5af305446726c378a30e53a445113682c768180c073bc042837e0b14
